repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go

```diff
@@ -0,0 +1,312 @@
+	"github.com/ipfs/go-ipfs/repo/fsrepo/migrations"
+	iface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/ipfs/interface-go-ipfs-core/options"
+	ipath "github.com/ipfs/interface-go-ipfs-core/path"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+)
+
+const (
+	// Default maximum download size
+	defaultFetchLimit = 1024 * 1024 * 512
+
+	tempNodeTcpAddr = "/ip4/127.0.0.1/tcp/0"
+)
+
+type IpfsFetcher struct {
+	distPath  string
+	limit     int64
+	bootstrap []string
+	peers     []peer.AddrInfo
+
+	openOnce  sync.Once
+	openErr   error
+	closeOnce sync.Once
+	closeErr  error
+
+	ipfs         iface.CoreAPI
+	ipfsTmpDir   string
+	ipfsStopFunc func()
+
+	fetched []ipath.Path
+	mutex   sync.Mutex
+
+	addrInfo peer.AddrInfo
+}
+
+// NewIpfsFetcher creates a new IpfsFetcher
+//
+// Specifying "" for distPath sets the default IPNS path.
+// Specifying 0 for fetchLimit sets the default, -1 means no limit.
+func NewIpfsFetcher(distPath string, fetchLimit int64, bootstrap []string, peers []peer.AddrInfo) *IpfsFetcher {
+	f := &IpfsFetcher{
+		limit:     defaultFetchLimit,
+		distPath:  migrations.LatestIpfsDist,
+		bootstrap: bootstrap,
+		peers:     peers,
+	}
+
+	if distPath != "" {
+		if !strings.HasPrefix(distPath, "/") {
+			distPath = "/" + distPath
+		}
+		f.distPath = distPath
+	}
+
+	if fetchLimit != 0 {
+		if fetchLimit < 0 {
+			fetchLimit = 0
+		}
+		f.limit = fetchLimit
+	}
+
+	return f
+}
+
+// Fetch attempts to fetch the file at the given path, from the distribution
+// site configured for this HttpFetcher.  Returns io.ReadCloser on success,
+// which caller must close.
+func (f *IpfsFetcher) Fetch(ctx context.Context, filePath string) (io.ReadCloser, error) {
+	// Initialize and start IPFS node on first call to Fetch, since the fetcher
+	// may be created by not used.
+	f.openOnce.Do(func() {
+		f.ipfsTmpDir, f.openErr = initTempNode(ctx, f.bootstrap, f.peers)
+		if f.openErr != nil {
+			return
+		}
+
+		f.openErr = f.startTempNode(ctx)
+	})
+
+	fmt.Printf("Fetching with IPFS: %q\n", filePath)
+
+	if f.openErr != nil {
+		return nil, f.openErr
+	}
+
+	iPath, err := parsePath(path.Join(f.distPath, filePath))
+	if err != nil {
+		return nil, err
+	}
+
+	nd, err := f.ipfs.Unixfs().Get(ctx, iPath)
+	if err != nil {
+		return nil, err
+	}
+
+	f.recordFetched(iPath)
+
+	fileNode, ok := nd.(files.File)
+	if !ok {
+		return nil, fmt.Errorf("%q is not a file", filePath)
+	}
+
+	if f.limit != 0 {
+		return migrations.NewLimitReadCloser(fileNode, f.limit), nil
+	}
+	return fileNode, nil
+}
+
+func (f *IpfsFetcher) Close() error {
+	f.closeOnce.Do(func() {
+		if f.ipfsStopFunc != nil {
+			// Tell ipfs node to stop and wait for it to stop
+			f.ipfsStopFunc()
+		}
+
+		if f.ipfsTmpDir != "" {
+			// Remove the temp ipfs dir
+			f.closeErr = os.RemoveAll(f.ipfsTmpDir)
+		}
+	})
+	return f.closeErr
+}
+
+func (f *IpfsFetcher) AddrInfo() peer.AddrInfo {
+	return f.addrInfo
+}
+
+func (f *IpfsFetcher) FetchedPaths() []ipath.Path {
+	return f.fetched
```

Comment:
This was meant to be without a mutex, since it must only be read after all fetching is complete.  I have added the mutex for consistency, but if it were needed here then that is really a misuse of this function.

Reply:
Makes sense, we may need to revisit this later